pkg/types/namespacedname.go

```diff
@@ -0,0 +1,58 @@
+/*
+Copyright 2015 Google Inc. All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package types
+
+// NamespacedName comprises a resource name, with a mandatory namespace,
+// rendered as "<namespace>/<name>".  Being a type captures intent and
+// helps make sure that UIDs, namespaced names and non-namespaced names
+// do not get conflated in code.  It also provides a convenient place for
+// utility functions to do things like splitting, joining and rendering.
+type NamespacedName struct {
+	namespace string
+	name      string
+}
+
+// NewNamespacedName is the constructor
+func NewNamespacedName(namespace, name string) *NamespacedName {
```

Comment:
I don't see any circular import - what am I missing?  I don't see why this warrants a new package.

The fact that this TINY util class now grows a dep on validation is wrong to me.  We do have a million places in tests that create invalid internal objects, but the fix is not to sprinkle validation  hither and thither.  We should fix it systemically, and until then, we should fix things up as they are needed.

Also, what Clayton said about name formats is true, albeit something I hope to fix eventually.

I could live with a simple len(input) != 0 sort of check - that will catch bad tests without depending on pkg/api, but the complexity leaks through.  If your "real" factory function returns value-or-error, then you have to check for error all over.  I am asserting this is has neutral-to-negative value in real code.

I'd rather see

```
// in the main .go file
NewNamespacedName(ns, name string) NamespacedName {
    return NamespacedName{ns, name}
}

// in the testing.go file, same pkg
NewNamespacedNameOrDie(ns, name string) NamespacedName {
    if len(ns) == 0 || len(name) == 0 {
        panic("invalid call to NewNamespacedNameOrDie(%q, %q)", ns, name)
    }
```

then call that is tests, and let "real code" assume that input has been verified already.